### PR TITLE
SignalFx exporter: Add host metadata synchronization

### PR DESCRIPTION
**Pull request by dmitryax**
**Description:** 
This change introduces host metadata updates functionality in Signalfx exporter. Synchronization happens once first metric batch is received. It looks for dimensions set by resourcedetection processor

### CODE REVIEW

**Comment by dmitryax** [comment]
as discussed with @jrcamp it will be changes to use cloud unique id instead of host

**Comment by jrcamp** [comment]
@dmitryax I think the PR has become too big. I think it'd be good to break it up at least into two pieces, one being stuff that came from smart agent that's already been vetted there and the second is new code that may need a closer look.

**Comment by dmitryax** [comment]
@jrcamp broke it down as you suggested. The first part: https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/1118

File: exporter/signalfxexporter/hostmetadata/metadata.go
Comment (dmitryax) [reply exporter/signalfxexporter/hostmetadata/metadata.go:82]: thanks. i'll do that using sync.Once

File: exporter/signalfxexporter/hostmetadata/metadata.go
Comment (dmitryax) [reply exporter/signalfxexporter/hostmetadata/metadata.go:36]: not anymore. good catch, thanks.

File: exporter/signalfxexporter/hostmetadata/metadata.go
Comment (dmitryax) [reply exporter/signalfxexporter/hostmetadata/metadata.go:27]: that makes sense. I'll update it in this PR

Comment (dmitryax) [reply exporter/signalfxexporter/hostmetadata/metadata.go]: updated

File: exporter/signalfxexporter/hostmetadata/metadata.go
Comment (dmitryax) [reply exporter/signalfxexporter/hostmetadata/metadata.go]: fixed

File: exporter/signalfxexporter/go.mod
Comment (jrcamp) [inline exporter/signalfxexporter/go.mod:18]: Use testify assert instead of adding new dep.

**Comment by jrcamp** [comment]
lgtm after the assert thing is fixed and #1100 is merged (should be soon) so that can be removed here.

File: exporter/signalfxexporter/go.mod
Comment (dmitryax) [reply exporter/signalfxexporter/go.mod:18]: oh nice catch! I got it automatically imported

Comment (jrcamp) [reply exporter/signalfxexporter/go.mod]: Yeah it's always importing the wrong one for me haha. Major pain.

**Comment by dmitryax** [comment]
@jrcamp , PR rebased

**Review by jrcamp — APPROVED** [review]

**Pull request merged**
